data_structures/src/transaction_factory.rs

```diff
@@ -121,23 +136,96 @@ pub trait OutputsCollection {
             )
             .ok_or_else(|| TransactionError::OutputValueOverflow)?;
 
-        let amount = output_value
-            .checked_add(fee)
-            .ok_or_else(|| TransactionError::OutputValueOverflow)?;
-
-        let (output_pointers, input_value) =
-            self.take_enough_utxos(amount, timestamp, block_number_limit, utxo_strategy)?;
-        let inputs: Vec<Input> = output_pointers.into_iter().map(Input::new).collect();
+        // For the first estimation: 1 input and 1 output more for the change address
+        let mut current_weight = calculate_weight(1, outputs.len() + 1, dr_output, max_weight)?;
+
+        let max_iterations = 1 + ((u64::from(max_weight) - current_weight) / u64::from(INPUT_SIZE));
+        for _i in 0..max_iterations {
+            let fee = match fee_type {
+                FeeType::Absolute => fee,
```

Comment:
I would expect the Absolute fee check to be outside of this for loop, to make this function easier to understand. But that means duplicating some code.